File: backend/schemas/changeset.py
"""ChangeSet 对外 DTO；前端只依赖这些字段，不读取业务表快照来推断状态。"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


ChangeSetStatusLiteral = Literal[
    "draft", "resolving", "proposed", "validating", "preflighting",
    "awaiting_approval", "committing", "applied", "cancelled", "rejected",
    "failed", "expired", "conflicted",
]


class ChangeCheckDTO(BaseModel):
    id: str
    change_set_id: str
    check_type: str
    check_key: str
    input: Dict[str, Any] = Field(default_factory=dict)
    result: Dict[str, Any] = Field(default_factory=dict)
    status: Literal["pending", "running", "passed", "failed", "skipped"]
    actor_id: Optional[str] = None
    actor_type: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime


class ChangeEventDTO(BaseModel):
    id: str
    change_set_id: str
    sequence: int
    event_type: str
    from_status: Optional[str] = None
    to_status: Optional[str] = None
    actor_id: Optional[str] = None
    actor_type: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class ChangeSetDTO(BaseModel):
    id: str
    org_id: str
    resource_type: str
    resource_id: str
    operation: str
    base_revision: str
    base_snapshot: Dict[str, Any] = Field(default_factory=dict)
    proposed_snapshot: Dict[str, Any] = Field(default_factory=dict)
    patch: List[Dict[str, Any]] = Field(default_factory=list)
    diff: Dict[str, Any] = Field(default_factory=dict)
    risk_level: Literal["low", "medium", "high", "critical"]
    policy_snapshot: Dict[str, Any] = Field(default_factory=dict)
    plan_snapshot: Optional[Dict[str, Any]] = None
    tool_policy_snapshot: Optional[Dict[str, Any]] = None
    check_summary: Optional[Dict[str, Any]] = None
    status: ChangeSetStatusLiteral
    idempotency_key: str
    expires_at: datetime
    created_by: str
    created_by_type: str
    updated_by: Optional[str] = None
    updated_by_type: Optional[str] = None
    audit_subject: Dict[str, Any] = Field(default_factory=dict)
    recovery_of_id: Optional[str] = None
    committed_revision: Optional[str] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    conflict: Optional[Dict[str, Any]] = None
    revision: int
    created_at: datetime
    updated_at: datetime
    checks: List[ChangeCheckDTO] = Field(default_factory=list)


class ChangeSetTimelineDTO(BaseModel):
    change_set_id: str
    events: List[ChangeEventDTO] = Field(default_factory=list)


class CancelChangeSetRequest(BaseModel):
    reason: str = Field(default="", max_length=500)


class RecoverChangeSetRequest(BaseModel):
    idempotency_key: Optional[str] = Field(default=None, min_length=1, max_length=200)

File: frontend/src/types/changeset.ts
/** ChangeSet 对外契约：聊天消息/表单是投影，不能替代这些状态与时间线。 */

export const CHANGESET_CONTRACT_VERSION = 'changeset.v1' as const;

export type ChangeSetStatus =
  | 'draft' | 'resolving' | 'proposed' | 'validating' | 'preflighting'
  | 'awaiting_approval' | 'committing' | 'applied'
  | 'cancelled' | 'rejected' | 'failed' | 'expired' | 'conflicted';

export type ChangeSetRiskLevel = 'low' | 'medium' | 'high' | 'critical';

export interface ChangeCheck {
  id: string;
  change_set_id: string;
  check_type: string;
  check_key: string;
  input: Record<string, unknown>;
  result: Record<string, unknown>;
  status: 'pending' | 'running' | 'passed' | 'failed' | 'skipped';
  actor_id?: string | null;
  actor_type?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
}

export interface ChangeEvent {
  id: string;
  change_set_id: string;
  sequence: number;
  event_type: string;
  from_status?: ChangeSetStatus | null;
  to_status?: ChangeSetStatus | null;
  actor_id?: string | null;
  actor_type?: string | null;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface ChangeSet {
  id: string;
  org_id: string;
  resource_type: string;
  resource_id: string;
  operation: string;
  base_revision: string;
  base_snapshot: Record<string, unknown>;
  proposed_snapshot: Record<string, unknown>;
  patch: Array<Record<string, unknown>>;
  diff: Record<string, unknown>;
  risk_level: ChangeSetRiskLevel;
  policy_snapshot: Record<string, unknown>;
  plan_snapshot?: Record<string, unknown> | null;
  tool_policy_snapshot?: Record<string, unknown> | null;
  check_summary?: Record<string, unknown> | null;
  status: ChangeSetStatus;
  idempotency_key: string;
  expires_at: string;
  created_by: string;
  created_by_type: string;
  updated_by?: string | null;
  updated_by_type?: string | null;
  audit_subject: Record<string, unknown>;
  recovery_of_id?: string | null;
  committed_revision?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  conflict?: Record<string, unknown> | null;
  revision: number;
  created_at: string;
  updated_at: string;
  checks: ChangeCheck[];
  /** 第二批稳定投影，第三批直接渲染，不需读取业务表拼接。 */
  risk?: Record<string, unknown> | null;
  plan?: Record<string, unknown> | null;
  approval_actions?: Array<{
    action: 'confirm' | 'reject' | 'cancel';
    enabled: boolean;
    method: string;
    path: string;
  }>;
  result?: Record<string, unknown> | null;
}

export interface ChangeSetTimeline {
  change_set_id: string;
  events: ChangeEvent[];
}

/**
 * 聊天消息只保存这个引用（以及可选的展示快照），不保存流程状态。
 * 该引用是 ChangeSetCard 的唯一查询入口。
 */
export interface ChangeSetMessageReference {
  type: 'changeset';
  change_set_id: string;
  title?: string;
  resource_type?: string;
  snapshot?: Record<string, unknown>;
}
